### ISSUE TITLE
Request: Support for awaitables in LineEditor custom key handlers

### ISSUE THREAD

**Issue by Menyadar** [ronf/asyncssh#705]
I ran into an issue when trying to work with `LineEditor`. Provided examples suggest using `await process.stdin.readline()` to process submitted lines. This works fine, however it fails when i need to process the line (and possibly adjust output) before it is echoed on the screen and user pastes multiple lines. This is because `SSHLineEditor.process_input()` writes all the data to the output before they are available for `process.stdin.readline()`. Let's say i need to prefix every line with something like shell caret: `> ` and call some methods on each line before moving to the next, possibly printing some extra output if given line is incorrect.

Paste buffer:
```
1
2
3
```

Expected behavior:
```
> 1
> 2
> 3
```

Result when calling `line = await process.stdin.readline()` in a while loop and drawing caret after processing the line:
```
1
2
3
> > > 
```

Moving away from `process.stdin.readline()` and using custom key handler for return character fixed this issue and generally looks like much cleaner solution to me, however custom handlers are synchronous and don't support awaitables.

Sample handler:

```python
    def handle_RETURN(self, line: str, pos: int) -> bool | tuple[str, int]:  # <-- this should be async
        # first caret was written on channel open
        self.channel._editor._end_line()  # noqa
        try:
            self.process_line(line)  # <-- need awaitable here
        except Exception as e:
            self.render_or_handle(e)
        self.channel.write(self.caret)
        return "", 0
```

**Comment by ronf** [comment]
Yes - custom key handlers would be a good choice here. Unfortunately, the entire line editor module is callback-based, using the Channel/Session framework, and by design none of those functions are asynchronous. In fact, for things to function properly each callback must complete in its entirety before returning to the event loop, and none of these callbacks should block in any way. So, it's not really possible for any of the code in the line editor to do an await.

For your specific case, I think you may be able to use the option `line_echo`.  It allows you to set whether or not to echo completed input lines when they are entered, rather than waiting for the application to read and echo them, defaulting to `True`. Setting this to `False` and performing the echo in the application can better synchronize input and output, especially when there are input prompts.

Basically, the line editor will erase your input line when you hit <Return>, but you can modify the input before you echo it. Also, you can output things like a prompt before calling `readline()` again and that prompt will be output before the next line of input is returned. So, when you echo that line in your application, it will be after the correct prompt.

**Comment by Menyadar** [comment]
Well, that is.. erm.. _unfortunate_ :( I have some prior experience with **Twisted**, where callbacks are typically structured as chainable deferred objects (awaitables) so i'm surprised to encounter a situation in which user-defined code within an asynchronous library could not be handled asynchronously.

Tinkering with `line_echo` could be the way, however it will make it difficult to mix that with key handlers, right? At least from my understanding key handlers are executed before any line is available to `readline()`.

**Comment by ronf** [comment]
Do you have any need for key handlers for anything other than <return>? If not, you should be able to use `line_echo` without involving any key handlers. The `readline()` call will return with the input line the user typed, but with that input line erased. You would then be free to modify it if you needed to, and then you could write that new input line to the output. This would also guarantee that each input line was output after any prompt you had written, even when pasting multiple lines at once. No key handlers would be required here, and you'd be free to make async calls from the function you were calling `readline()` from.

As for not supporting async calls, part of that is historical -- the original version of this code actually pre-dates asyncio even existing in Python. It goes back to an older module called asyncore. This was before the "await" and "async" syntax even existed (and before "yield from" that predated "await"). Even now with modern asyncio, there are two levels of APIs available. The low-level API is what I'm using here, and it involves Transport and Protocol classes which define callbacks that are triggered when I/O occurs on the sockets. Callbacks can be chained as you said, but it is done with direct calls, and there are no coroutines, so you can't suspend a callback waiting for some other operation to complete.

It would probably be possible to write a version of the line editor which used the high-level asyncio API, and still keep most of the existing code intact. However, the run-to-completion nature of the calls mean that you'd have to suspend processing new input coming in while a prior key handler was blocked, or else input would be processed out of order. Some kind of queue would probably need to be involved.

**Comment by ronf** [comment]
Here's an example of using line_echo to deal with the prompt problem:

```python
import asyncio, asyncssh, sys

async def handle_session(stdin, stdout, stderr):
    total = 0
    prompt = '> '

    try:
        stdout.write('Enter numbers:\n')

        while True:
            stdout.write(prompt)
            line = await stdin.readline()
            if not line:
                break

            # Modify line or change prompt here if you need to
            stdout.write(line)

            try:
                total += int(line)
            except ValueError:
                stdout.write('Invalid number: %s' % line)
    except asyncssh.BreakReceived:
        pass

    stdout.write('\rTotal = %s\n' % total)
    stdout.channel.exit(0)

async def start_server():
    await asyncssh.listen('', 8022, server_host_keys=['ssh_host_key'],
                          authorized_client_keys='ssh_user_ca',
                          session_factory=handle_session, line_echo=False)
```

**Comment by Menyadar** [comment]
I have few key handlers registered, but it seems i could get rid of them when using `readline()`. The only one i'm unsure with is one for `INT` / `^C`, but it's pretty basic. I'll check if it'll work with these changes without custom hacks.

For the sync callbacks, i'd have to dig deeper to your code to see where the switch happens, as i assume there is a place where you have to await on some async part. Still trying to mentally match that to my experiences with Twisted.

**Comment by ronf** [comment]
The Ctrl-C can be handled by the BreakReceived in my example above. Also, hitting Ctrl-D at the beginning of a line is treated as EOF, and will cause `readline()` to return an empty string, which is also handled in the above example.

You can still use other key handlers if you need them -- they just need to be synchronous.

Regarding callbacks, there's no switch here -- the line editor inserts itself as an alternative to the original SSHServerChannel that was opened, and that was already receiving callbacks from the SSHServerConnection when data messages arrive. That in turn was being driven by callbacks coming from a call to asyncio.create_server(). There are no coroutines or await calls in any of that code. See the example at https://docs.python.org/3/library/asyncio-protocol.html#tcp-echo-server for a simple example of this.

**Comment by Menyadar** [comment]
Thanks. I'll try to move to custom `session_factory`. I'll need to have at least one key handler (`TAB`), but that can be synchronous and if it's possible to mix these two approaches, it should be ok.

**Comment by ronf** [comment]
You don't have to use session_factory here -- you could also use process_factory, which may be a better choice depending on what features you want when operating on the channel. What are you using now?

**Comment by Menyadar** [comment]
Right now `SSHServerProcess` from a custom `SSHServer.session_requested()`, but it's easy to modify your proposal to match `SSHServerProcess`. I used it just because you described it as a _to-go_ solution for new code in some previous issue.

**Comment by ronf** [comment]
Yeah -- if you want to use an SSHServerProcess, just replace the session_factory in my example with process_factory and a function that takes an SSHServerProcess as an argument. You can also do session_requested() if you like, but that's only needed if you want to be selective about which channel requests you want to allow, or select between different handlers depending on things like the user or client address.

**Comment by Menyadar** [comment]
> that's only needed if you want to be selective about which channel requests you want to allow, or select between different handlers 

this ^^, as i need to update connection options on some conditions and it's quite convenient to keep this stuff in `SSHServer` subclass.

**Comment by Menyadar** [comment]
Thanks for your time (and all your _code_!), i think i have all that i need to continue.

**Comment by ronf** [comment]
Yup - you can stick with that and still use my example above. Just have the handler you return to take a process argument and change references to stdin/stdout to process.stdin/process.stdout, and a few other things can reference process methods like `exit()` rather than going through `stdout.channel`.

**Comment by ronf** [comment]
Happy to help!
